### PR TITLE
client/rpcserver: Add cancel route.

### PR DESCRIPTION
**Pull request by JoeGruffins**

### CODE REVIEW

File: client/rpcserver/rpcserver.go [hunk 1]
```diff
@@ -52,16 +52,17 @@ var (
...
 	CreateWallet(appPass, walletPass []byte, form *core.WalletForm) error
 	Exchanges() (exchanges map[string]*core.Exchange)
 	InitializeClient(appPass []byte) error
 	Login(appPass []byte) (*core.LoginResult, error)
-	OpenWallet(assetID uint32, pw []byte) error
+	OpenWallet(assetID uint32, appPass []byte) error
 	GetFee(addr, cert string) (fee uint64, err error)
 	Register(form *core.RegisterForm) error
-	Sync(dex string, base, quote uint32) (*core.OrderBook, *core.BookFeed, error)
+	Sync(host string, base, quote uint32) (*core.OrderBook, *core.BookFeed, error)
```
Comment (chappjc) [inline client/rpcserver/rpcserver.go:65]: dex-> host is fine, but TBH I was fine with dex.  Just as long as it's no longer url.  :)

File: client/rpcserver/types.go [hunk 2]
```diff
@@ -12,6 +12,9 @@ import (
 	"decred.org/dcrdex/dex/encode"
 )
 
+// An orderID is a 256 bit number encoded as a hex string.
+const orderIdLen = 64
```
Comment (chappjc) [inline client/rpcserver/types.go:16]: Let's use `hex.EncodedLen(decred.org/dcrdex/dex/order.OrderIDSize)` or just `2*order.OrderIDSize` unless importing dex/order would be a problem.

File: client/rpcserver/types.go [hunk 3]
```diff
@@ -42,7 +45,7 @@ type getFeeResponse struct {
 
 // tradeResponse is used when responding to the trade route.
 type tradeResponse struct {
-	OrderID string `json:"orderid"`
+	OrderID string `json:"orderID"`
```
Comment (chappjc) [inline client/rpcserver/types.go:48]: I don't know what passes for idiomatic json, but I though we were using all lower.  Leaving that up to @buck54321

File: client/rpcserver/types.go [hunk 4]
```diff
@@ -68,11 +71,18 @@ type helpForm struct {
 	IncludePasswords bool   `json:"includepasswords"`
 }
 
+// tradeForm is information necessary to trade.
```
Comment (chappjc) [inline client/rpcserver/types.go:74]: Almost too vague.  How about "tradeForm combines the application password and the user's trade details" or something like that.  The docs you get from your IDE might as well let you know the intent of the type.

File: client/rpcserver/types.go [hunk 5]
```diff
@@ -270,3 +280,22 @@ func parseTradeArgs(params *RawParams) (*tradeForm, error) {
...
+		return nil, fmt.Errorf("%w: orderID has incorrect length", errArgs)
+	}
+	// We can only check up to 64 bits at a time. Break the 256 bit hex
+	// into 4 sections and check each.
+	div := orderIdLen / 4
+	for i := 0; i < 4; i++ {
+		if _, err := checkUIntArg(id[i*div:(i+1)*div], "orderID", 16, 64); err != nil {
+			return nil, fmt.Errorf("%w: invalid order id hex", errArgs)
+		}
+	}
```
Comment (chappjc) [inline client/rpcserver/types.go:299]: I'd be way more comfortable with `_, err := hex.DecodeString(id)` and then just replicate error message formatting here.

Comment (chappjc) [reply client/rpcserver/types.go:299]: Note that `dex/order.IDFromHex` does both, although it has an unnecessary `copy` for our case.

File: client/rpcserver/types.go (hunk 3 above)
Comment (buck54321) [reply client/rpcserver/types.go:48]: I've been inconsistent again. This is fine.

Comment (JoeGruffins) [reply client/rpcserver/types.go:48]: @buck54321 I don't have an opinion, but I would like them to be the same for the same values. An all lower-case rule would make that simpler however. Let me know

File: client/rpcserver/types.go [hunk 6]
```diff
@@ -289,13 +291,8 @@ func parseCancelArgs(params *RawParams) (*cancelForm, error) {
...
 	}
-	// We can only check up to 64 bits at a time. Break the 256 bit hex
-	// into 4 sections and check each.
-	div := orderIdLen / 4
-	for i := 0; i < 4; i++ {
-		if _, err := checkUIntArg(id[i*div:(i+1)*div], "orderID", 16, 64); err != nil {
-			return nil, fmt.Errorf("%w: invalid order id hex", errArgs)
-		}
+	if _, err := hex.DecodeString(id); err != nil {
+		return nil, fmt.Errorf("%w: invalid order id hex", errArgs)
```
Comment (chappjc) [inline client/rpcserver/types.go:295]: Wow, `%w` is great: https://blog.golang.org/go1.13-errors#TOC_3.3.  I didn't know about this addition to the language.

**Comment by chappjc** [comment]
As expected:

```
 $    ./dexcctl -u u -P p trade 127.0.0.1:7232 true false 42 0 100000000 10000 0
App password:
{
  "orderID": "4e96d67c36200813c351f1d564f7b93310bc6ec12b65254d2aeca155d26eaf26",
  "sig": "3044022036c4cc2e1cf1c1e30e66e7df9cfcac397060bd49fbe7bc3e35a32051f25c83a002203c3a1cb73f6c6c38704e60db6e616a5903ec2de3a60ea7e9a6aed7892c139c48",
  "stamp": 1591131274731
}

$    ./dexcctl -u u -P p help cancel
cancel "orderID"

Cancel an order.

Args:
    orderID (string): The hex ID of the order to cancel

Returns:
    string: The message "canceled order [order ID]"

$    ./dexcctl -u u -P p cancel 4e96d67c36200813c351f1d564f7b93310bc6ec12b65254d2aeca155d26eaf26
App password:
canceled order 4e96d67c36200813c351f1d564f7b93310bc6ec12b65254d2aeca155d26eaf26
```

```
[DBG] CORE: notify: |POKE| (order) Order placed - buying 1.00000000 dcr (4e96d67c) - Order: 4e96d67c36200813c351f1d564f7b93310bc6ec12b65254d2aeca155d26eaf26
[TRC] CORE: notify: |DATA| (balance) balance updated
[DBG] CORE: notify: |POKE| (order) Preimage sent - match cycle has begun for order 4e96d67c - Order: 4e96d67c36200813c351f1d564f7b93310bc6ec12b65254d2aeca155d26eaf26
[DBG] RPC: authenticated user with ip: 127.0.0.1:56330
[DBG] RPC: authenticated user with ip: 127.0.0.1:56362
[DBG] CORE: notify: |POKE| (order) Preimage sent - match cycle has begun for cancellation order for trade 4e96d67c - Order: 
[INF] CORE: maker notification for cancel order received for order 4e96d67c36200813c351f1d564f7b93310bc6ec12b65254d2aeca155d26eaf26. match id = baacd20842e743d64edf1a98dbe14b7e0e19a470f51e2511cd116bafea88654f
[INF] CORE: notify: |SUCCESS| (order) Order canceled - Buy order on dcr-btc at 127.0.0.1:7232 has been canceled (4e96d67c) - Order: 4e96d67c36200813c351f1d564f7b93310bc6ec12b65254d2aeca155d26eaf26
[TRC] CORE: notify: |DATA| (order) cancel - Order: 
[INF] CORE: taker notification for cancel order received for order ea1ff20d610c7647f16a5c2b1409f775cdcba8d6e71023f4b0f625d73cbbfe72. match id = baacd20842e743d64edf1a98dbe14b7e0e19a470f51e2511cd116bafea88654f
```

**Pull request merged**